Add unit-testing for executorpodfactory

foxish opened the pull request:
This is important for unit-testing executorpodfactory, which will be part of our first upstream PR.

cc @apache-spark-on-k8s/contributors @liyinan926

foxish committed:
Unit test for executorpodfactory

foxish committed:
Fix test

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+  private val executorPrefix: String = "base"
+  private val executorImage: String = "executor-image"
+  private val driverPod = new PodBuilder()
+    .withNewMetadata()
```
Indentation

foxish replied:
Done

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+      .set(KUBERNETES_DRIVER_POD_NAME, driverPodName)
+      .set(KUBERNETES_EXECUTOR_POD_NAME_PREFIX, executorPrefix)
+      .set(EXECUTOR_DOCKER_IMAGE, executorImage)
+    sc = new SparkContext("local", "test")
```
Can we use mocks here? Can we use a mock SecurityManager?

foxish replied:
Would that be useful here?
The `ExecutorPodFactoryImpl` class doesn't use SecurityManager.

mccheah replied:
More just that we don't want to make a real SparkContext here. `SparkContext.clearActiveContext` could be problematic on parallel test runs. If we don't need to make a SparkContext, we should avoid it.

foxish committed:
Indentation fix

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+    // Default memory limit is 1024M + 384M (minimum overhead constant).
+    assert(executor.getSpec.getContainers.size() == 1)
+    assert(executor.getSpec.getContainers.get(0).getImage == executorImage)
+    assert(executor.getSpec.getContainers.get(0).getVolumeMounts.size() == 0)
```
Use `isEmpty`.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+      getLimits.get("memory").getAmount == "1408Mi")
+
+    // The pod has no node selector, volumes.
+    assert(executor.getSpec.getNodeSelector.size() == 0)
```
`isEmpty` - replace this everywhere.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+  test("basic executor pod has reasonable defaults") {
+    val factory = new ExecutorPodFactoryImpl(baseConf,
+      NodeAffinityExecutorPodModifierImpl, None, None, None, None, None)
+    val executor = factory.createExecutorPod("1", "dummy", "dummy",
```
Move all args to the next line and all on one line if they fit in 100 chars. If not - put each arg on its own line.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+
+    val factory = new ExecutorPodFactoryImpl(conf,
+      NodeAffinityExecutorPodModifierImpl, None, None, None, None, None)
+    val executor = factory.createExecutorPod("1",
```
Move args to one per line.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+    assert(executor.getSpec.getContainers.size() == 1)
+    assert(executor.getSpec.getContainers.get(0).getVolumeMounts.size() == 1)
+    assert(executor.getSpec.getContainers.get(0).getVolumeMounts.get(0).getName == "secret1-volume")
+    assert(executor.getSpec.getContainers.get(0).getVolumeMounts.get(0).
```
There's a couple of places where the dot is on the end of the line followed by the method or field being on the next line. I think we want to move the dot to the next line and continue the chain of calls from there.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+      SparkTransportConf.fromSparkConf(conf, "shuffle"),
+      sc.env.securityManager,
+      sc.env.securityManager.isAuthenticationEnabled())
+    val shuffleManager = new KubernetesExternalShuffleManagerImpl(conf,
```
Don't use the impl - use a mock.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+    val shuffleManager = new KubernetesExternalShuffleManagerImpl(conf,
+      kubernetesClient, kubernetesExternalShuffleClient)
+    val factory = new ExecutorPodFactoryImpl(conf,
+      NodeAffinityExecutorPodModifierImpl, None, None, None, None, Some(shuffleManager))
```
Don't use the impl, use a mock.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+      kubernetesClient, kubernetesExternalShuffleClient)
+    val factory = new ExecutorPodFactoryImpl(conf,
+      NodeAffinityExecutorPodModifierImpl, None, None, None, None, Some(shuffleManager))
+    val executor = factory.createExecutorPod("1",
```
Argument formatting as stated elsewhere in the review - fix this in all applicable locations.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+
+    assert(executor.getSpec.getContainers.size() == 1)
+    assert(executor.getSpec.getContainers.get(0).getVolumeMounts.size() == 1)
+    assert(executor.getSpec.getContainers.get(0).getVolumeMounts.get(0).getName == "0-tmp")
```
nit - all assertions should use triple-equals - `===` - instead of double equals.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+      NodeAffinityExecutorPodModifierImpl, None, Some(smallFiles), None, None, None)
+    val executor = factory.createExecutorPod("1",
+      "dummy", "dummy", Seq[(String, String)](), driverPod, Map[String, Int]())
+
```
Extra whitespace here

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,228 @@
+  }
+
+  // Check that the expected environment variables are present.
+  private def checkEnv(executor: Pod, additionalEnvVars: Set[String]): Unit = {
```
Do we not also want to be checking the values of the environment variables we are setting, not just the keys? Hence the `additionalEnvVars` should be a `Map`, not a `Set`.

foxish committed:
Fix isEmpty and split between lines

foxish committed:
Address issues with multi-line code fragments

foxish committed:
Replace == with ===

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,265 @@
+  test("init-container bootstrap step adds an init container") {
+    val conf = baseConf.clone()
+
+    val initContainerBootstrap = new SparkPodInitContainerBootstrapImpl(
```
Use a mock and not the impl. Add a separate unit test for this class if it does not already exist.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,265 @@
+    conf.set(KUBERNETES_SHUFFLE_NAMESPACE, "default")
+    conf.set(KUBERNETES_SHUFFLE_DIR, "/tmp")
+
+    val kubernetesExternalShuffleClient = new KubernetesExternalShuffleClientImpl(
```
Use a mock, not the impl.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,265 @@
+      SparkTransportConf.fromSparkConf(conf, "shuffle"),
+      sc.env.securityManager,
+      sc.env.securityManager.isAuthenticationEnabled())
+    val shuffleManager = new KubernetesExternalShuffleManagerImpl(
```
Use a mock, not an impl.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,265 @@
+
+    val factory = new ExecutorPodFactoryImpl(
+      conf,
+      NodeAffinityExecutorPodModifierImpl,
```
Use a mock, not the impl. Add a separate unit test for this impl class.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,265 @@
+    val secretsBootstrap = new MountSecretsBootstrapImpl(Map("secret1" -> "/var/secret1"))
+    val factory = new ExecutorPodFactoryImpl(
+      conf,
+      NodeAffinityExecutorPodModifierImpl,
```
Use a mock

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,265 @@
+  test("basic executor pod has reasonable defaults") {
+    val factory = new ExecutorPodFactoryImpl(
+      baseConf,
+      NodeAffinityExecutorPodModifierImpl,
```
Use a mock - replace this everywhere

erikerlandson committed:
mock shuffleManager

erikerlandson commented:
Does `NodeAffinityExecutorPodModifierImpl` object need to be mocked? The problem seemed to be around use of `sc` as dep for construction of `shuffleManager`

erikerlandson commented:
CI build failing?
```
[error] /home/jenkins/workspace/PR-spark-k8s-full-build/resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactorySuite.scala:34: object kubernetes is not a member of package org.apache.spark.deploy
[error] import org.apache.spark.deploy.kubernetes.{constants, SparkPodInitContainerBootstrapImpl}
[error]                                ^
```

Trying to figure out if I broke something here, but I didn't change that import, and my edits were all local to the unit test file.

mccheah commented:
The packages have been renamed to `k8s` right?

mccheah commented:
> Does NodeAffinityExecutorPodModifierImpl object need to be mocked? The problem seemed to be around use of sc as dep for construction of shuffleManager

Basically we only want to have the impl for `ExecutorPodFactory` - `SparkContext` and any other object should be mocks.

erikerlandson commented:
I rebased this branch to the `v2.2.0-kubernetes-0.4.0` release tag to pick up the `k8s` rename and the other remaining 0.4.0 commits

erikerlandson committed:
.kubernetes. => .k8s.

erikerlandson committed:
move to k8s subdir

erikerlandson committed:
fix package clause to k8s

erikerlandson committed:
mock nodeAffinityExecutorPodModifier

erikerlandson commented:
@foxish I just rebased this to the 0.4.0 tag, to pick up the `.k8s.` rename - however this PR is currently targeted to branch-2.2.-kubernetes. Was the intent to merge it to branch-2.2.-kubernetes and then port it to #498 ?

foxish commented:
Yes we should merge this into the 2.2 branch first; only a subset of the
tests are valid for the first apache/spark PR.

On Thu, Sep 28, 2017 at 3:27 PM, Erik Erlandson <notifications@github.com>
wrote:

> @foxish <https://github.com/foxish> I just rebased this to the 0.4.0 tag,
> to pick up the .k8s. rename - however this PR is currently targeted to
> branch-2.2.-kubernetes. Was the intent to merge it to
> branch-2.2.-kubernetes and then port it to #498
> <https://github.com/apache-spark-on-k8s/spark/pull/498> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache-spark-on-k8s/spark/pull/491#issuecomment-332981135>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3U5_586I5BPRTbzgvppAvvDQIFY4Vgks5snB1hgaJpZM4PYWCO>
> .
>



-- 
Anirudh Ramanathan

erikerlandson commented:
OK I'll make sure its rebased to 2.2 branch before it merges

erikerlandson committed:
remove commented code

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,286 @@
+  private var baseConf: SparkConf = _
+
+  private val nodeAffinityExecutorPodModifier = mock(classOf[NodeAffinityExecutorPodModifier])
+  when(nodeAffinityExecutorPodModifier.addNodeAffinityAnnotationIfUseful(
```
Move this to the `before` block.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,286 @@
+    any(classOf[Map[String, Int]]))).thenAnswer(AdditionalAnswers.returnsFirstArg())
+
+  before {
+    SparkContext.clearActiveContext()
```
Don't think we need to do this anymore.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,286 @@
+  test("init-container bootstrap step adds an init container") {
+    val conf = baseConf.clone()
+
+    val initContainerBootstrap = new SparkPodInitContainerBootstrapImpl(
```
Don't use the impl, use a mock.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,286 @@
+
+  test("Small-files add a secret & secret volume mount to the container") {
+    val conf = baseConf.clone()
+    val smallFiles = new MountSmallFilesBootstrapImpl("secret1", "/var/secret1")
```
Don't use the impl, use a mock

erikerlandson committed:
move when clause to before{} block

erikerlandson committed:
mock initContainerBootstrap, smallFiles

erikerlandson committed:
insert actual logic into smallFiles mock

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,325 @@
+    val setEnvs = executor.getSpec.getContainers.get(0).getEnv.asScala.map {
+      x => x.getName
+    }.toSet
+    assert(defaultEnvs === setEnvs)
```
Do we want to check the values for these as well?

erikerlandson replied:
It's not obvious to me what the expected values for some of the env-vars ought to be. Possibly @foxish knows. I've been assuming the main logic being tested there is purely env propagation - to confirm what vars are being passed under what circumstances.

foxish replied:
Done

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,325 @@
+      .set(EXECUTOR_DOCKER_IMAGE, executorImage)
+    when(nodeAffinityExecutorPodModifier.addNodeAffinityAnnotationIfUseful(
+      any(classOf[Pod]),
+      any(classOf[Map[String, Int]]))).thenAnswer(AdditionalAnswers.returnsFirstArg())
```
This setup makes it such that we don't check if `nodeAffinityExecutorPodModifier` is used in the pod factory. Either:
1. Use `Mockito.verify`
2. Return meaningful modifications to the pod/container from the mock and test that such modifications were done in the appropriate locations.

erikerlandson committed:
verify application of nodeAffinityExecutorPodModifier

erikerlandson committed:
avoid cumulative invocation

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactorySuite.scala:
```diff
@@ -0,0 +1,349 @@
+          val container = invocation.getArgumentAt(1, classOf[Container])
+          val secretName = "secret1"
+          val secretMountPath = "/var/secret1"
+          val resolvedPod = new PodBuilder(pod)
```
These can be simplified. Just return the pod with an extra label/annotation, and the container with an extra environment variable. Verify that the pod has this extra label/annotation and the container has the appropriate environment variable.

Or, use `Mockito.verify()` - I'm still not sure which is the stronger/more idiomatic approach here.

foxish replied:
I don't like that we're mocking the very behavior we're testing here. I'd actually be in favor of omitting mocks for this particular case entirely and just relying on a real stamped out `MountSmallFilesBootstrapImpol` since it's not nearly complex enough to need a mock IMO.

mccheah commented:
@erikerlandson @foxish ping on latest comment?

foxish committed:
Fixed env-var check to include values, removed mock for small files

foxish commented:
Addressed comment - like I mentioned there, I see the value in mocks elsewhere, but not in that particular case, since using the actual class itself seems very straightforward. Also addressed comment about env-variable value checking in addition to just the key names. This should be ready to merge at this point. PTAL @mccheah.

mccheah reviewed (approved):
Might take another pass over later to remove the use of `MountSecretsBootstrapImpl`, but I want to move this along.
